File: LogisimPlus/LogisimPlus/logisim/tools/move/Connector.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System;
using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.tools.move
{

	using ReplacementMap = logisim.circuit.ReplacementMap;
	using Wire = logisim.circuit.Wire;
	using Direction = logisim.data.Direction;
	using Location = logisim.data.Location;

	internal class Connector
	{
		private const int MAX_SECONDS = 10;
		private const int MAX_ORDERING_TRIES = 10;
		private const int MAX_SEARCH_ITERATIONS = 20000;

		private Connector()
		{
		}

		internal const string ALLOW_NEITHER = "neither";
		internal const string ALLOW_VERTICAL = "vert";
		internal const string ALLOW_HORIZONTAL = "horz";

		internal static MoveResult computeWires(MoveRequest req)
		{
			MoveGesture gesture = req.MoveGesture;
			int dx = req.DeltaX;
			int dy = req.DeltaY;
			List<ConnectionData> baseConnects;
			baseConnects = new List<ConnectionData>(gesture.Connections);
			List<ConnectionData> impossible = pruneImpossible(baseConnects, gesture.FixedAvoidanceMap, dx, dy);

			AvoidanceMap selAvoid = AvoidanceMap.create(gesture.Selected, dx, dy);
			Dictionary<ConnectionData, ISet<Location>> pathLocs;
			pathLocs = new Dictionary<ConnectionData, ISet<Location>>();
			Dictionary<ConnectionData, IList<SearchNode>> initNodes;
			initNodes = new Dictionary<ConnectionData, IList<SearchNode>>();
			foreach (ConnectionData conn in baseConnects)
			{
				HashSet<Location> connLocs = new HashSet<Location>();
				List<SearchNode> connNodes = new List<SearchNode>();
				processConnection(conn, dx, dy, connLocs, connNodes, selAvoid);
				pathLocs[conn] = connLocs;
				initNodes[conn] = connNodes;
			}

			MoveResult bestResult = null;
			int tries;
			switch (baseConnects.Count)
			{
			case 0:
				tries = 0;
				break;
			case 1:
				tries = 1;
				break;
			case 2:
				tries = 2;
				break;
			case 3:
				tries = 8;
				break;
			default:
				tries = MAX_ORDERING_TRIES;
			break;
			}
			long stopTime = DateTimeHelper.CurrentUnixTimeMillis() + MAX_SECONDS * 1000;
			for (int tryNum = 0; tryNum < tries && stopTime - DateTimeHelper.CurrentUnixTimeMillis() > 0; tryNum++)
			{
				if (ConnectorThread.OverrideRequested)
				{
					return null;
				}
				List<ConnectionData> connects;
				connects = new List<ConnectionData>(baseConnects);
				if (tryNum < 2)
				{
					sortConnects(connects, dx, dy);
					if (tryNum == 1)
					{
						connects.Reverse();
					}
				}
				else
				{
					Collections.shuffle(connects);
				}

				MoveResult candidate = tryList(req, gesture, connects, dx, dy, pathLocs, initNodes, stopTime);
				if (candidate == null)
				{
					return null;
				}
				else if (bestResult == null)
				{
					bestResult = candidate;
				}
				else
				{
					int unsatisfied1 = bestResult.UnsatisifiedConnections.Count;
					int unsatisfied2 = candidate.UnsatisifiedConnections.Count;
					if (unsatisfied2 < unsatisfied1)
					{
						bestResult = candidate;
					}
					else if (unsatisfied2 == unsatisfied1)
					{
						int dist1 = bestResult.TotalDistance;
						int dist2 = candidate.TotalDistance;
						if (dist2 < dist1)
						{
							bestResult = candidate;
						}
					}
				}
			}
			if (bestResult == null)
			{ // should only happen for no connections
				bestResult = new MoveResult(req, new ReplacementMap(), impossible, 0);
			}
			else
			{
				bestResult.addUnsatisfiedConnections(impossible);
			}
			return bestResult;
		}

		private static List<ConnectionData> pruneImpossible(List<ConnectionData> connects, AvoidanceMap avoid, int dx, int dy)
		{
			List<Wire> pathWires = new List<Wire>();
			foreach (ConnectionData conn in connects)
			{
				foreach (Wire w in conn.WirePath)
				{
					pathWires.Add(w);
				}
			}

			List<ConnectionData> impossible = new List<ConnectionData>();
			for (IEnumerator<ConnectionData> it = connects.GetEnumerator(); it.MoveNext();)
			{
				ConnectionData conn = it.Current;
				Location dest = conn.Location.translate(dx, dy);
				if (avoid.get(dest) != null)
				{
					bool isInPath = false;
					foreach (Wire w in pathWires)
					{
						if (w.contains(dest))
						{
							isInPath = true;
							break;
						}
					}
					if (!isInPath)
					{
// JAVA TO C# CONVERTER TASK: .NET enumerators are read-only:
						it.remove();
						impossible.Add(conn);
					}
				}
			}
			return impossible;
		}

		/// <summary>
		/// Creates a list of the connections to make, sorted according to their location. If, for example, we are moving an
		/// east-facing AND gate southeast, then we prefer to connect the inputs from the top down to minimize the chances
		/// that the created wires will interfere with each other - but if we are moving that gate northeast, we prefer to
		/// connect the inputs from the bottom up.
		/// </summary>
		private static void sortConnects(List<ConnectionData> connects, in int dx, in int dy)
		{
			connects.Sort(new ComparatorAnonymousInnerClass(dx, dy));
		}

		private class ComparatorAnonymousInnerClass : IComparer<ConnectionData>
		{
			private int dx;
			private int dy;

			public ComparatorAnonymousInnerClass(int dx, int dy)
			{
				this.dx = dx;
				this.dy = dy;
			}

			public int Compare(ConnectionData ac, ConnectionData bc)
			{
				Location a = ac.Location;
				Location b = bc.Location;
				int abx = a.X - b.X;
				int aby = a.Y - b.Y;
				return abx * dx + aby * dy;
			}
		}

		private static void processConnection(ConnectionData conn, int dx, int dy, HashSet<Location> connLocs, List<SearchNode> connNodes, AvoidanceMap selAvoid)
		{
			Location cur = conn.Location;
			Location dest = cur.translate(dx, dy);
			if (selAvoid.get(cur) == null)
			{
				Direction preferred = conn.Direction;
				if (preferred == null)
				{
					if (Math.Abs(dx) > Math.Abs(dy))
					{
						preferred = dx > 0 ? Direction.East : Direction.West;
					}
					else
					{
						preferred = dy > 0 ? Direction.South : Direction.North;
					}
				}

				connLocs.Add(cur);
				connNodes.Add(new SearchNode(conn, cur, preferred, dest));
			}

			foreach (Wire w in conn.WirePath)
			{
				foreach (Location loc in w)
				{
					if (selAvoid.get(loc) == null || loc.Equals(dest))
					{
						bool added = connLocs.Add(loc);
						if (added)
						{
							Direction dir = null;
							if (w.endsAt(loc))
							{
								if (w.Vertical)
								{
									int y0 = loc.Y;
									int y1 = w.getOtherEnd(loc).Y;
									dir = y0 < y1 ? Direction.North : Direction.South;
								}
								else
								{
									int x0 = loc.X;
									int x1 = w.getOtherEnd(loc).X;
									dir = x0 < x1 ? Direction.West : Direction.East;
								}
							}
							connNodes.Add(new SearchNode(conn, loc, dir, dest));
						}
					}
				}
			}
		}

		private static MoveResult tryList(MoveRequest req, MoveGesture gesture, List<ConnectionData> connects, int dx, int dy, Dictionary<ConnectionData, ISet<Location>> pathLocs, Dictionary<ConnectionData, IList<SearchNode>> initNodes, long stopTime)
		{
			AvoidanceMap avoid = gesture.FixedAvoidanceMap.cloneMap();
			avoid.markAll(gesture.Selected, dx, dy);

			ReplacementMap replacements = new ReplacementMap();
			List<ConnectionData> unconnected = new List<ConnectionData>();
			int totalDistance = 0;
			foreach (ConnectionData conn in connects)
			{
				if (ConnectorThread.OverrideRequested)
				{
					return null;
				}
				if (DateTimeHelper.CurrentUnixTimeMillis() - stopTime > 0)
				{
					unconnected.Add(conn);
					continue;
				}
				IList<SearchNode> connNodes = initNodes[conn];
				ISet<Location> connPathLocs = pathLocs[conn];
				SearchNode n = findShortestPath(connNodes, connPathLocs, avoid);
				if (n != null)
				{ // normal case - a path was found
					totalDistance += n.Distance;
					List<Location> path = convertToPath(n);
					processPath(path, conn, avoid, replacements, connPathLocs);
				}
				else if (ConnectorThread.OverrideRequested)
				{
					return null; // search was aborted: return null to indicate this
				}
				else
				{
					unconnected.Add(conn);
				}
			}
			return new MoveResult(req, replacements, unconnected, totalDistance);
		}

		private static SearchNode findShortestPath(IList<SearchNode> nodes, ISet<Location> pathLocs, AvoidanceMap avoid)
		{
			PriorityQueue<SearchNode> q = new PriorityQueue<SearchNode>(nodes);
			HashSet<SearchNode> visited = new HashSet<SearchNode>();
			int iters = 0;
			while (!q.isEmpty() && iters < MAX_SEARCH_ITERATIONS)
			{
				iters++;
				SearchNode n = q.remove();
				if (iters % 64 == 0 && ConnectorThread.OverrideRequested || n == null)
				{
					return null;
				}
				if (n.Destination)
				{
					return n;
				}
				bool added = visited.Add(n);
				if (!added)
				{
					continue;
				}
				Location loc = n.Location;
				Direction dir = n.Direction;
				int neighbors = 3;
				object allowed = avoid.get(loc);
				if (allowed != null && n.Start && pathLocs.Contains(loc))
				{
					allowed = null;
				}
				if (allowed == ALLOW_NEITHER)
				{
					neighbors = 0;
				}
				else if (allowed == ALLOW_VERTICAL)
				{
					if (dir == null)
					{
						dir = Direction.North;
						neighbors = 2;
					}
					else if (dir == Direction.North || dir == Direction.South)
					{
						neighbors = 1;
					}
					else
					{
						neighbors = 0;
					}
				}
				else if (allowed == ALLOW_HORIZONTAL)
				{
					if (dir == null)
					{
						dir = Direction.East;
						neighbors = 2;
					}
					else if (dir == Direction.East || dir == Direction.West)
					{
						neighbors = 1;
					}
					else
					{
						neighbors = 0;
					}
				}
				else
				{
					if (dir == null)
					{
						dir = Direction.North;
						neighbors = 4;
					}
					else
					{
						neighbors = 3;
					}
				}
				for (int i = 0; i < neighbors; i++)
				{
					Direction oDir;
					switch (i)
					{
					case 0:
						oDir = dir;
						break;
					case 1:
						oDir = neighbors == 2 ? dir.reverse() : dir.rotateCCW();
						break;
					case 2:
						oDir = dir.rotateCW();
						break;
					default: // must be 3
						oDir = dir.reverse();
					break;
					}
					SearchNode o = n.next(oDir, allowed != null);
					if (o != null && !visited.Contains(o))
					{
						q.add(o);
					}
				}
			}
			return null;
		}

		private static List<Location> convertToPath(SearchNode last)
		{
			SearchNode next = last;
			SearchNode prev = last.Previous;
			List<Location> ret = new List<Location>();
			ret.Add(next.Location);
			while (prev != null)
			{
				if (prev.Direction != next.Direction)
				{
					ret.Add(prev.Location);
				}
				next = prev;
				prev = prev.Previous;
			}
			if (!ret[ret.Count - 1].Equals(next.Location))
			{
				ret.Add(next.Location);
			}
			ret.Reverse();
			return ret;
		}

		private static void processPath(List<Location> path, ConnectionData conn, AvoidanceMap avoid, ReplacementMap repl, ISet<Location> unmarkable)
		{
			IEnumerator<Location> pathIt = path.GetEnumerator();
// JAVA TO C# CONVERTER TASK: Java iterators are only converted within the context of 'while' and 'for' loops:
			Location loc0 = pathIt.next();
			if (!loc0.Equals(conn.Location))
			{
				Location pathLoc = conn.WirePathStart;
				bool found = loc0.Equals(pathLoc);
				foreach (Wire w in conn.WirePath)
				{
					Location nextLoc = w.getOtherEnd(pathLoc);
					if (found)
					{ // existing wire will be removed
						repl.remove(w);
						avoid.unmarkWire(w, nextLoc, unmarkable);
					}
					else if (w.contains(loc0))
					{ // wires after this will be removed
						found = true;
						if (!loc0.Equals(nextLoc))
						{
							avoid.unmarkWire(w, nextLoc, unmarkable);
							Wire shortenedWire = Wire.create(pathLoc, loc0);
							repl.replace(w, shortenedWire);
							avoid.markWire(shortenedWire, 0, 0);
						}
					}
					pathLoc = nextLoc;
				}
			}
			while (pathIt.MoveNext())
			{
				Location loc1 = pathIt.Current;
				Wire newWire = Wire.create(loc0, loc1);
				repl.add(newWire);
				avoid.markWire(newWire, 0, 0);
				loc0 = loc1;
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/tools/move/Connector.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.tools.move;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.PriorityQueue;
import java.util.Set;

import logisim.circuit.ReplacementMap;
import logisim.circuit.Wire;
import logisim.data.Direction;
import logisim.data.Location;

class Connector {
	private static final int MAX_SECONDS = 10;
	private static final int MAX_ORDERING_TRIES = 10;
	private static final int MAX_SEARCH_ITERATIONS = 20000;

	private Connector() {
	}

	static final String ALLOW_NEITHER = "neither";
	static final String ALLOW_VERTICAL = "vert";
	static final String ALLOW_HORIZONTAL = "horz";

	static MoveResult computeWires(MoveRequest req) {
		MoveGesture gesture = req.getMoveGesture();
		int dx = req.getDeltaX();
		int dy = req.getDeltaY();
		ArrayList<ConnectionData> baseConnects = new ArrayList<>(gesture.getConnections());
		ArrayList<ConnectionData> impossible = pruneImpossible(baseConnects, gesture.getFixedAvoidanceMap(), dx, dy);

		AvoidanceMap selAvoid = AvoidanceMap.create(gesture.getSelected(), dx, dy);
		HashMap<ConnectionData, Set<Location>> pathLocs = new HashMap<>();
		HashMap<ConnectionData, List<SearchNode>> initNodes = new HashMap<>();
		for (ConnectionData conn : baseConnects) {
			HashSet<Location> connLocs = new HashSet<>();
			ArrayList<SearchNode> connNodes = new ArrayList<>();
			processConnection(conn, dx, dy, connLocs, connNodes, selAvoid);
			pathLocs.put(conn, connLocs);
			initNodes.put(conn, connNodes);
		}

		MoveResult bestResult = null;
		int tries = switch (baseConnects.size()) {
			case 0 -> 0;
			case 1 -> 1;
			case 2 -> 2;
			case 3 -> 8;
			default -> MAX_ORDERING_TRIES;
		};
		long stopTime = System.currentTimeMillis() + MAX_SECONDS * 1000;
		for (int tryNum = 0; tryNum < tries && stopTime - System.currentTimeMillis() > 0; tryNum++) {
			if (ConnectorThread.isOverrideRequested()) return null;
			ArrayList<ConnectionData> connects = new ArrayList<>(baseConnects);
			if (tryNum < 2) {
				sortConnects(connects, dx, dy);
				if (tryNum == 1) Collections.reverse(connects);
			} else Collections.shuffle(connects);

			MoveResult candidate = tryList(gesture, connects, dx, dy, pathLocs, initNodes, stopTime);
			if (candidate == null) return null;
			else if (bestResult == null) bestResult = candidate;
			else {
				int unsatisfied1 = bestResult.getUnsatisifiedConnections().size();
				int unsatisfied2 = candidate.getUnsatisifiedConnections().size();
				if (unsatisfied2 < unsatisfied1) bestResult = candidate;
				else if (unsatisfied2 == unsatisfied1) {
					int dist1 = bestResult.getTotalDistance();
					int dist2 = candidate.getTotalDistance();
					if (dist2 < dist1) bestResult = candidate;
				}
			}
		}
		// should only happen for no connections
		if (bestResult == null) return new MoveResult(new ReplacementMap(), impossible, 0);
		else bestResult.addUnsatisfiedConnections(impossible);
		return bestResult;
	}

	private static ArrayList<ConnectionData> pruneImpossible(ArrayList<ConnectionData> connects, AvoidanceMap avoid,
			int dx, int dy) {
		ArrayList<Wire> pathWires = new ArrayList<>();
		for (ConnectionData conn : connects) pathWires.addAll(conn.getWirePath());

		ArrayList<ConnectionData> impossible = new ArrayList<>();
		for (Iterator<ConnectionData> it = connects.iterator(); it.hasNext();) {
			ConnectionData conn = it.next();
			Location dest = conn.getLocation().add(dx, dy);
			if (avoid.get(dest) != null) {
				boolean isInPath = false;
				for (Wire w : pathWires)
					if (w.contains(dest)) {
						isInPath = true;
						break;
					}
				if (!isInPath) {
					it.remove();
					impossible.add(conn);
				}
			}
		}
		return impossible;
	}

	/**
	 * Creates a list of the connections to make, sorted according to their location. If, for example, we are moving an
	 * east-facing AND gate southeast, then we prefer to connect the inputs from the top down to minimize the chances
	 * that the created wires will interfere with each other - but if we are moving that gate northeast, we prefer to
	 * connect the inputs from the bottom up.
	 */
	private static void sortConnects(ArrayList<ConnectionData> connects, final int dx, final int dy) {
		connects.sort((ac, bc) -> {
			Location a = ac.getLocation();
			Location b = bc.getLocation();
			int abx = a.x() - b.x();
			int aby = a.y() - b.y();
			return abx * dx + aby * dy;
		});
	}

	private static void processConnection(ConnectionData conn, int dx, int dy, HashSet<Location> connLocs,
			ArrayList<SearchNode> connNodes, AvoidanceMap selAvoid) {
		Location cur = conn.getLocation();
		Location dest = cur.add(dx, dy);
		if (selAvoid.get(cur) == null) {
			Direction preferred = conn.getDirection();
			if (preferred == null)
				if (Math.abs(dx) > Math.abs(dy)) preferred = dx > 0 ? Direction.East : Direction.West;
				else preferred = dy > 0 ? Direction.South : Direction.North;

			connLocs.add(cur);
			connNodes.add(new SearchNode(conn, cur, preferred, dest));
		}

		for (Wire w : conn.getWirePath())
			for (Location loc : w)
				if (selAvoid.get(loc) == null || loc.equals(dest)) {
					boolean added = connLocs.add(loc);
					if (added) {
						Direction dir = null;
						if (w.endsAt(loc)) if (w.isVertical()) {
							int y0 = loc.y();
							int y1 = w.getOtherEnd(loc).y();
							dir = y0 < y1 ? Direction.North : Direction.South;
						}
						else {
							int x0 = loc.x();
							int x1 = w.getOtherEnd(loc).x();
							dir = x0 < x1 ? Direction.West : Direction.East;
						}
						connNodes.add(new SearchNode(conn, loc, dir, dest));
					}
				}
	}

	private static MoveResult tryList(MoveGesture gesture, ArrayList<ConnectionData> connects, int dx,
									  int dy, HashMap<ConnectionData, Set<Location>> pathLocs,
									  HashMap<ConnectionData, List<SearchNode>> initNodes, long stopTime) {
		AvoidanceMap avoid = gesture.getFixedAvoidanceMap().cloneMap();
		avoid.markAll(gesture.getSelected(), dx, dy);

		ReplacementMap replacements = new ReplacementMap();
		ArrayList<ConnectionData> unconnected = new ArrayList<>();
		int totalDistance = 0;
		for (ConnectionData conn : connects) {
			if (ConnectorThread.isOverrideRequested()) return null;
			if (System.currentTimeMillis() - stopTime > 0) {
				unconnected.add(conn);
				continue;
			}
			List<SearchNode> connNodes = initNodes.get(conn);
			Set<Location> connPathLocs = pathLocs.get(conn);
			SearchNode n = findShortestPath(connNodes, connPathLocs, avoid);
			if (n != null) { // normal case - a path was found
				totalDistance += n.getDistance();
				ArrayList<Location> path = convertToPath(n);
				processPath(path, conn, avoid, replacements, connPathLocs);
			} else if (ConnectorThread.isOverrideRequested())
				return null; // search was aborted: return null to indicate this
			else unconnected.add(conn);
		}
		return new MoveResult(replacements, unconnected, totalDistance);
	}

	private static SearchNode findShortestPath(List<SearchNode> nodes, Set<Location> pathLocs, AvoidanceMap avoid) {
		PriorityQueue<SearchNode> q = new PriorityQueue<>(nodes);
		HashSet<SearchNode> visited = new HashSet<>();
		int iters = 0;
		while (!q.isEmpty() && iters < MAX_SEARCH_ITERATIONS) {
			iters++;
			SearchNode n = q.remove();
			if (iters % 64 == 0 && ConnectorThread.isOverrideRequested() || n == null) return null;
			if (n.isDestination()) return n;
			boolean added = visited.add(n);
			if (!added) continue;
			Location loc = n.getLocation();
			Direction dir = n.getDirection();
			int neighbors;
			String allowed = (String) avoid.get(loc);
			if (allowed != null && n.isStart() && pathLocs.contains(loc)) allowed = null;
			switch (allowed) {
				case ALLOW_NEITHER -> neighbors = 0;
				case ALLOW_VERTICAL -> {
					if (dir == null) {
						dir = Direction.North;
						neighbors = 2;
					} else if (dir == Direction.North || dir == Direction.South) neighbors = 1;
					else neighbors = 0;
				}
				case ALLOW_HORIZONTAL -> {
					if (dir == null) {
						dir = Direction.East;
						neighbors = 2;
					} else if (dir == Direction.East || dir == Direction.West) neighbors = 1;
					else neighbors = 0;
				}
				case null, default -> {
					if (dir == null) {
						dir = Direction.North;
						neighbors = 4;
					} else neighbors = 3;
				}
			}
			for (int i = 0; i < neighbors; i++) {
				Direction oDir = switch (i) {
					case 0 -> dir;
					case 1 -> neighbors == 2 ? dir.reverse() : dir.rotateCCW();
					case 2 -> dir.rotateCW();
					default -> // must be 3
							dir.reverse();
				};
				SearchNode o = n.next(oDir, allowed != null);
				if (o != null && !visited.contains(o)) q.add(o);
			}
		}
		return null;
	}

	private static ArrayList<Location> convertToPath(SearchNode last) {
		SearchNode next = last;
		SearchNode prev = last.getPrevious();
		ArrayList<Location> ret = new ArrayList<>();
		ret.add(next.getLocation());
		while (prev != null) {
			if (prev.getDirection() != next.getDirection()) ret.add(prev.getLocation());
			next = prev;
			prev = prev.getPrevious();
		}
		if (!ret.getLast().equals(next.getLocation())) ret.add(next.getLocation());
		Collections.reverse(ret);
		return ret;
	}

	private static void processPath(ArrayList<Location> path, ConnectionData conn, AvoidanceMap avoid,
			ReplacementMap repl, Set<Location> unmarkable) {
		Iterator<Location> pathIt = path.iterator();
		Location loc0 = pathIt.next();
		if (!loc0.equals(conn.getLocation())) {
			Location pathLoc = conn.getWirePathStart();
			boolean found = loc0.equals(pathLoc);
			for (Wire w : conn.getWirePath()) {
				Location nextLoc = w.getOtherEnd(pathLoc);
				if (found) { // existing wire will be removed
					repl.remove(w);
					avoid.unmarkWire(w, nextLoc, unmarkable);
				} else if (w.contains(loc0)) { // wires after this will be removed
					found = true;
					if (!loc0.equals(nextLoc)) {
						avoid.unmarkWire(w, nextLoc, unmarkable);
						Wire shortenedWire = Wire.create(pathLoc, loc0);
						repl.replace(w, shortenedWire);
						avoid.markWire(shortenedWire, 0, 0);
					}
				}
				pathLoc = nextLoc;
			}
		}
		while (pathIt.hasNext()) {
			Location loc1 = pathIt.next();
			Wire newWire = Wire.create(loc0, loc1);
			repl.add(newWire);
			avoid.markWire(newWire, 0, 0);
			loc0 = loc1;
		}
	}
}
